docs/RFCS/20171214_computed_columns.md

```diff
@@ -0,0 +1,155 @@
+- Start Date: 2017-12-14
+- Authors: Daniel Harrison, Justin Jaffray
+- RFC PR:
+- Cockroach Issue:
+
+# Summary
+
+Computed columns allow users to create columns which are not directly writable,
+but are instead pure functions of other column values in that row. This is
+supported in several major databases, including MySQL, Oracle, and SQL Server.
+This RFC covers both the version of computed columns that are physically stored
+alongside other columns (referred to here as *materialized*) and computed
+columns which are only computed when read (*non-materialized*).
+
+# Motivation
+
+Computed columns allow storing data derived from other columns without putting
+the burden of keeping that data updated on the client. See [Guide Level
+Explanation: Partitioning](#partitioning) and [Guide Level Explanation: JSON](#json) for the two
+motivating examples.
+
+
+# Guide Level Explanation
+
+The grammar for a computed column is `column_name <type> AS <expr>
+[PERSISTED]`, where `<expr>` is a pure function of non-computed columns in the
+same table. `PERSISTED` indicates a materialized computed column and is
+required if that columns appears in the primary key.
+
+## Partitioning
+
+[Partitioning](https://github.com/cockroachdb/cockroach/blob/aa61db043e9c54c0b83a405cd76ce0ec7cc6a35d/docs/RFCS/20170921_sql_partitioning.md)
+requires that partitions are defined using columns that are a prefix of the
+primary key. In the case of geo-partitioning, some applications will want to
+collapse the number of possible values in this columns, to make certain classes
+of queries more performant (see [Query planning changes in the
+RFC](https://github.com/cockroachdb/cockroach/blob/aa61db043e9c54c0b83a405cd76ce0ec7cc6a35d/docs/RFCS/20170921_sql_partitioning.md#query-planning-changes)
+for details). For example, if a `users` table has a `country` and `state`
+column, then a CockroachDB operator could make a materialized computed column
+`locality` with a reduced domain for use in partitioning.
+
+```sql
+CREATE TABLE users (
+  locality STRING AS CASE
+    WHEN country IN ('ca', 'mx', 'us') THEN 'north_america'
+    WHEN country IN ('au', 'nz') THEN 'australia'
+  END PERSISTED,
+  id SERIAL,
+  name STRING,
+  country STRING,
+  ...
+  PRIMARY KEY (locality, id)
+) PARTITION BY LIST (locality) (
+  PARTITION north_america VALUES IN ('north_america'),
+  PARTITION australia VALUES IN ('australia')
+)
+```
+
+## JSON
+
+When the primary source of truth for a table lives in a JSON blob, it can be desirable to put an index on a particular field of a JSON document. In particular, computed columns allow for the following use-case: a two column table, with a primary key column and a payload column, whose primary key is computed as some field from the payload column. This alleviates the need for the client to manually separate their JSON blobs from their primary keys.
+
+```sql
+CREATE TABLE documents (
+  id STRING PRIMARY KEY AS payload->>'id' PERSISTED,
+  payload JSONB
+)
```

Comment:
users might prefer using the postgres variant 

```
CREATE TABLE documents (
   payload JSONB
   UNIQUE (payload->>'id')
 )
```

I see the JSON use case as not necessarily needing the computed column but a computed index.